src/lib/apply-patch/src/lib.rs

```diff
@@ -0,0 +1,218 @@
+            let found_range = if self.ignore_line_numbers {
+                // todo: implement line-number-agnostic patching. mvp: all our patch files have exact line numbers known.
+                unimplemented!()
+            } else {
+                self.find_fixed_range(original_content, hunk, &expected_lines)?
+            };
+            // first copy to the output all the content between either the last patched range or the beginning of the file, and the beginning of this patch
+            patched_content.push_str(&original_content[patched_up_to..found_range.start]);
+            patched_up_to = found_range.end;
+            // then interleave the context lines with the added lines
+            self.patch_content(&original_content[found_range], hunk, &mut patched_content);
+        }
+        // finally, copy everything between the final patch and the end of the file
+        patched_content.push_str(&original_content[patched_up_to..]);
+        Ok(patched_content)
+    }
+
+    fn patch_content(&self, original_content: &str, hunk: &Hunk, patched_content: &mut String) {
+        let mut orig_lines = original_content.lines();
+        for hunk_line in &hunk.lines {
+            match *hunk_line {
+                Line::Context(_) => {
+                    // in the case of a context line, push the original line, not the one from the patch file
+                    // this may be a whitespace-insensitive patch, and we don't want to modify any lines that aren't marked `-`
+                    patched_content.push_str(orig_lines.next().unwrap());
+                    patched_content.push('\n');
+                }
+                Line::Add(s) => {
+                    patched_content.push_str(s);
+                    patched_content.push('\n');
+                }
+                Line::Remove(_) => {
+                    orig_lines.next().unwrap();
+                }
+            }
+        }
+    }
+
+    fn find_fixed_range(
+        &self,
+        original_content: &str,
+        hunk: &Hunk,
+        expected_lines: &[&str],
+    ) -> Result<Range<usize>, MismatchError> {
+        let mut line_indices = original_content
+            .match_indices('\n')
+            .map(|(newline_index, _)| newline_index + 1);
+        // line numbers are all 1-indexed - this is specifically the parsed number
+        let start_line = hunk.old_range.start as usize - 1;
+        // line n starts one past the n-1th newline
+        let start = if start_line == 0 {
+            0
+        } else {
+            line_indices
+                .nth(start_line - 1)
+                .ok_or_else(|| MismatchError::NotEnoughLines {
+                    expected: start_line - 1,
+                    found: original_content.lines().count(),
+                })?
+        };
+        // start..end should be the byte range in `content` to be patched
+        let end = line_indices.nth(expected_lines.len() - 1).ok_or_else(|| {
```

Comment:
Should this be the following? This is in part why I was hoping for more DRY, though I don't feel too strongly about these 4(8) LOC.

```suggestion
        let end = line_indices.nth(start_line + expected_lines.len() - 1).ok_or_else(|| {
```

Reply:
No. `nth` is stateful.

Reply:
Okay, got it now.  The earlier version had this, which made me wonder:
```
.nth(start_line -1)
...
expected: start_line - 1
```
followed by
```
.nth(expected_lines.len() - 1)
...
expected: expected_lines.len() - 1,
```

Reply:
Yes, it was a bug I caught while updating it. Only affected the error message.